File: backend/src/FarmManagement.Application/DTOs/Crops/CropLifecycleTemplateModels.cs
namespace FarmManagement.Application.DTOs.Crops;

public sealed record CropLifecycleStageResponse(
    Guid Id,
    Guid LifecycleTemplateId,
    string StageName,
    int SequenceNumber,
    int? ExpectedDurationDays,
    string? Description,
    bool IsActive);

public sealed record CropLifecycleTemplateResponse(
    Guid Id,
    Guid? OrganizationId,
    Guid CropId,
    string CropCode,
    string CropName,
    string Name,
    string? Description,
    bool IsDefault,
    bool IsSystem,
    bool IsActive,
    DateTimeOffset CreatedAt,
    Guid? CreatedBy,
    DateTimeOffset? UpdatedAt,
    Guid? UpdatedBy,
    IReadOnlyList<CropLifecycleStageResponse> Stages);

public sealed record CreateCropLifecycleTemplateRequest(
    Guid? CropId,
    string? Name,
    string? Description,
    bool IsDefault = false);

public sealed record UpdateCropLifecycleTemplateRequest(
    Guid? CropId,
    string? Name,
    string? Description,
    bool IsDefault = false);

public sealed record CreateCropLifecycleStageRequest(
    string? StageName,
    int SequenceNumber,
    int? ExpectedDurationDays,
    string? Description);

public sealed record UpdateCropLifecycleStageRequest(
    string? StageName,
    int SequenceNumber,
    int? ExpectedDurationDays,
    string? Description);

File: frontend/farm-management-web/src/app/core/crop-lifecycle-templates/crop-lifecycle-template.models.ts
import { PagedResponse } from "../models/paged-response.model";

export interface CropLifecycleStage {
  id: string;
  lifecycleTemplateId: string;
  stageName: string;
  sequenceNumber: number;
  expectedDurationDays: number | null;
  description: string | null;
  isActive: boolean;
}

export interface CropLifecycleTemplate {
  id: string;
  organizationId: string | null;
  cropId: string;
  cropName: string;
  name: string;
  description: string | null;
  isDefault: boolean;
  isSystem: boolean;
  isActive: boolean;
  createdAt: string;
  createdBy: string | null;
  updatedAt: string | null;
  updatedBy: string | null;
  stages: readonly CropLifecycleStage[];
}

export type CropLifecycleTemplateList = PagedResponse<CropLifecycleTemplate>;

export interface CropLifecycleTemplateQuery {
  page?: number;
  pageSize?: number;
  cropId?: string | null;
  isActive?: boolean | null;
}

export interface CreateCropLifecycleStageRequest {
  stageName: string;
  sequenceNumber: number;
  expectedDurationDays?: number | null;
  description?: string | null;
}

export interface UpdateCropLifecycleStageRequest {
  stageName: string;
  sequenceNumber: number;
  expectedDurationDays?: number | null;
  description?: string | null;
}

export interface CreateCropLifecycleTemplateRequest {
  cropId: string;
  name: string;
  description?: string | null;
  isDefault?: boolean;
  stages?: readonly CreateCropLifecycleStageRequest[];
}

export interface UpdateCropLifecycleTemplateRequest {
  cropId: string;
  name: string;
  description?: string | null;
  isDefault?: boolean;
}

export interface ReorderCropLifecycleStageItem {
  stageId: string;
  sequenceNumber: number;
}

export interface ReorderCropLifecycleStagesRequest {
  stages: readonly ReorderCropLifecycleStageItem[];
}
